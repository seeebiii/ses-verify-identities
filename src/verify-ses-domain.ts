import { CfnOutput, Fn } from 'aws-cdk-lib';
import { CnameRecord, HostedZone, IHostedZone, MxRecord, TxtRecord } from 'aws-cdk-lib/aws-route53';
import { Topic } from 'aws-cdk-lib/aws-sns';
import { AwsCustomResource, PhysicalResourceId } from 'aws-cdk-lib/custom-resources';
import { EnvironmentPlaceholders } from 'aws-cdk-lib/cx-api';
import { Construct } from 'constructs';
import { generateSesPolicyForCustomResource } from './helper';

export type NotificationType = 'Bounce' | 'Complaint' | 'Delivery';

/**
 * @struct
 */
export interface IVerifySesDomainProps {
  /**
   * A domain name to be used for the SES domain identity, e.g. 'sub-domain.example.org'
   */
  readonly domainName: string;
  /**
   * A hostedZone name to be matched with Route 53 record. e.g. 'example.org'
   * @default same as domainName
   */
  readonly hostedZoneName?: string;
  /**
   * Whether to automatically add a TXT record to the hosed zone of your domain. This only works if your domain is managed by Route53. Otherwise disable it.
   * @default true
   */
  readonly addTxtRecord?: boolean;
  /**
   * Whether to automatically add a MX record to the hosted zone of your domain. This only works if your domain is managed by Route53. Otherwise disable it.
   * @default true
   */
  readonly addMxRecord?: boolean;
  /**
   * Whether to automatically add DKIM records to the hosted zone of your domain. This only works if your domain is managed by Route53. Otherwise disable it.
   * @default true
   */
  readonly addDkimRecords?: boolean;
  /**
   * An SNS topic where bounces, complaints or delivery notifications can be sent to. If none is provided, a new topic will be created and used for all different notification types.
   * @default new topic will be created
   */
  readonly notificationTopic?: Topic;
  /**
   * Select for which notification types you want to configure a topic.
   * @default [Bounce, Complaint]
   */
  readonly notificationTypes?: NotificationType[];
  /**
   * Whether to include the original email headers in the notifications.
   * @default false
   */
  readonly includeOriginalHeaders?: boolean;
}

/**
 * A construct to verify a SES domain identity. It initiates a domain verification and can automatically create appropriate records in Route53 to verify the domain. Also, it's possible to attach a notification topic for bounces, complaints or delivery notifications.
 *
 * @example
 *
 * new VerifySesDomain(this, 'SesDomainVerification', {
 *   domainName: 'example.org'
 * });
 *
 */
export class VerifySesDomain extends Construct {
  constructor(parent: Construct, name: string, props: IVerifySesDomainProps) {
    super(parent, name);

    const domainName = props.domainName;
    const verifyDomainIdentity = this.verifyDomainIdentity(domainName);
    const topic = this.createTopicOrUseExisting(domainName, verifyDomainIdentity, props.notificationTopic);
    this.addTopicToDomainIdentity(domainName, topic, props.notificationTypes, props.includeOriginalHeaders);

    const hostedZoneName = props.hostedZoneName ? props.hostedZoneName : domainName;
    const zone = this.getHostedZone(hostedZoneName);

    if (isTrueOrUndefined(props.addTxtRecord)) {
      const txtRecord = this.createTxtRecordLinkingToSes(zone, domainName, verifyDomainIdentity);
      txtRecord.node.addDependency(verifyDomainIdentity);
    }

    if (isTrueOrUndefined(props.addMxRecord)) {
      const mxRecord = this.createMxRecord(zone, domainName);
      mxRecord.node.addDependency(verifyDomainIdentity);
    }

    if (isTrueOrUndefined(props.addDkimRecords)) {
      const verifyDomainDkim = this.initDkimVerification(domainName);
      verifyDomainDkim.node.addDependency(verifyDomainIdentity);
      this.addDkimRecords(verifyDomainDkim, zone, domainName);
    }
  }

  private verifyDomainIdentity(domainName: string): AwsCustomResource {
    return new AwsCustomResource(this, 'VerifyDomainIdentity', {
      onCreate: {
        service: 'SES',
        action: 'verifyDomainIdentity',
        parameters: {
          Domain: domainName,
        },
        physicalResourceId: PhysicalResourceId.fromResponse('VerificationToken'),
      },
      onUpdate: {
        service: 'SES',
        action: 'verifyDomainIdentity',
        parameters: {
          Domain: domainName,
        },
        physicalResourceId: PhysicalResourceId.fromResponse('VerificationToken'),
      },
      onDelete: {
        service: 'SES',
        action: 'deleteIdentity',
        parameters: {
          Identity: domainName,
        },
      },
      policy: generateSesPolicyForCustomResource('VerifyDomainIdentity', 'DeleteIdentity'),
    });
  }

  getHostedZone(domainName: string): IHostedZone {
    return HostedZone.fromLookup(this, 'Zone', {
      domainName: domainName,
    });
  }

  private createTxtRecordLinkingToSes(zone: IHostedZone, domainName: string, verifyDomainIdentity: AwsCustomResource) {
    return new TxtRecord(this, 'SesVerificationRecord', {
      zone,
      recordName: `_amazonses.${domainName}`,
      values: [verifyDomainIdentity.getResponseField('VerificationToken')],
    });
  }

  private createMxRecord(zone: IHostedZone, domainName: string) {
    return new MxRecord(this, 'SesMxRecord', {
      zone,
      recordName: domainName,
      values: [
        {
          hostName: Fn.sub(`inbound-smtp.${EnvironmentPlaceholders.CURRENT_REGION}.amazonaws.com`),
          priority: 10,
        },
      ],
    });
  }

  private initDkimVerification(domainName: string) {
    return new AwsCustomResource(this, 'VerifyDomainDkim', {
      onCreate: {
        service: 'SES',
        action: 'verifyDomainDkim',
        parameters: {
          Domain: domainName,
        },
        physicalResourceId: PhysicalResourceId.of(domainName + '-verify-domain-dkim'),
      },
      onUpdate: {
        service: 'SES',
        action: 'verifyDomainDkim',
        parameters: {
          Domain: domainName,
        },
        physicalResourceId: PhysicalResourceId.of(domainName + '-verify-domain-dkim'),
      },
      policy: generateSesPolicyForCustomResource('VerifyDomainDkim'),
    });
  }

  private addDkimRecords(verifyDomainDkim: AwsCustomResource, zone: IHostedZone, domainName: string) {
    [0, 1, 2].forEach((val) => {
      const dkimToken = verifyDomainDkim.getResponseField(`DkimTokens.${val}`);
      const cnameRecord = new CnameRecord(this, 'SesDkimVerificationRecord' + val, {
        zone,
        recordName: `${dkimToken}._domainkey.${domainName}`,
        domainName: `${dkimToken}.dkim.amazonses.com`,
      });
      cnameRecord.node.addDependency(verifyDomainDkim);
    });
  }

  private createTopicOrUseExisting(domainName: string, verifyDomainIdentity: AwsCustomResource, existingTopic?: Topic) {
    const topic = existingTopic ?? new Topic(this, 'SesNotificationTopic');
    new CfnOutput(this, domainName + 'SesNotificationTopic', {
      value: topic.topicArn,
      description: 'SES notification topic for ' + domainName,
    });
    topic.node.addDependency(verifyDomainIdentity);
    return topic;
  }

  private addTopicToDomainIdentity(domainName: string, topic: Topic, notificationTypes?: NotificationType[], includeOriginalHeaders?: boolean) {
    if (notificationTypes?.length) {
      notificationTypes.forEach((type) => {
        this.addSesNotificationTopicForIdentity(domainName, type, topic, includeOriginalHeaders);
      });
    } else {
      // by default, notify on errors like a bounce or complaint
      this.addSesNotificationTopicForIdentity(domainName, 'Bounce', topic, includeOriginalHeaders);
      this.addSesNotificationTopicForIdentity(domainName, 'Complaint', topic, includeOriginalHeaders);
    }
  }

  private addSesNotificationTopicForIdentity(
    identity: string,
    notificationType: NotificationType,
    notificationTopic: Topic,
    includeOriginalHeaders?: boolean,
  ): void {
    const addTopic = new AwsCustomResource(this, `Add${notificationType}Topic-${identity}`, {
      onCreate: {
        service: 'SES',
        action: 'setIdentityNotificationTopic',
        parameters: {
          Identity: identity,
          NotificationType: notificationType,
          SnsTopic: notificationTopic.topicArn,
        },
        physicalResourceId: PhysicalResourceId.of(`${identity}-set-${notificationType}-topic`),
      },
      policy: generateSesPolicyForCustomResource('SetIdentityNotificationTopic'),
    });

    addTopic.node.addDependency(notificationTopic);

    if (includeOriginalHeaders === true) {
      const addTopicHeaders = new AwsCustomResource(this, `Add${notificationType}TopicHeaders-${identity}`, {
        onCreate: {
          service: 'SES',
          action: 'setIdentityHeadersInNotificationsEnabled',
          parameters: {
            Enabled: true,
            Identity: identity,
            NotificationType: notificationType,
          },
          physicalResourceId: PhysicalResourceId.of(`${identity}-set-${notificationType}-topic-headers`),
        },
        policy: generateSesPolicyForCustomResource('SetIdentityHeadersInNotificationsEnabled'),
      });

      addTopicHeaders.node.addDependency(addTopic);
    }
  }
}

function isTrueOrUndefined(prop?: boolean): boolean {
  return prop === undefined || prop;
}
