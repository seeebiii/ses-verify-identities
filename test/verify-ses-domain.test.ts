
import * as cdk from 'aws-cdk-lib';
import { Capture, Match, Template } from 'aws-cdk-lib/assertions';
import { Topic } from 'aws-cdk-lib/aws-sns';
import { VerifySesDomain } from '../src';

const domain = 'sub-domain.example.org';
const hostedZoneName = 'example.org';
const hostedZoneId = '12345';
const zoneName = domain + '.';
VerifySesDomain.prototype.getHostedZone = jest.fn().mockReturnValue({
  HostedZoneId: hostedZoneId,
  zoneName: hostedZoneName,
});

describe('SES domain verification', () => {
  it('ensure custom resource exists to initiate domain verification', () => {
    const app = new cdk.App();
    const stack = new cdk.Stack(app, 'TestStack');

    new VerifySesDomain(stack, 'VerifyExampleDomain', {
      domainName: domain,
      hostedZoneName,
      addTxtRecord: false,
      addMxRecord: false,
      addDkimRecords: false,
    });

    // Then
    const template = Template.fromStack(stack);
    template.resourceCountIs('Custom::AWS', 3);
    template.hasResourceProperties('Custom::AWS', {
      Create: Match.serializedJson(Match.objectLike({
        service: 'SES',
        action: 'verifyDomainIdentity',
        parameters: {
          Domain: domain,
        },
      })),
      Delete: Match.serializedJson(Match.objectLike({
        service: 'SES',
        action: 'deleteIdentity',
        parameters: {
          Identity: domain,
        },
      })),
    },
    );
    template.resourceCountIs('AWS::SNS::Topic', 1);
    template.resourceCountIs('AWS::Route53::RecordSet', 0);
  });

  it('ensure txt record is added', () => {
    const app = new cdk.App();
    const stack = new cdk.Stack(app, 'TestStack');

    new VerifySesDomain(stack, 'VerifyExampleDomain', {
      domainName: domain,
      addTxtRecord: true,
      addMxRecord: false,
      addDkimRecords: false,
    });

    // Then
    const template = Template.fromStack(stack);
    template.resourceCountIs('Custom::AWS', 3);
    template.resourceCountIs('AWS::SNS::Topic', 1);
    template.resourceCountIs('AWS::Route53::RecordSet', 1);
    template.hasResourceProperties('AWS::Route53::RecordSet', {
      Type: 'TXT',
      Name: '_amazonses.' + zoneName,
    },
    );
  });

  it('ensure mx record is added', () => {
    const app = new cdk.App();
    const stack = new cdk.Stack(app, 'TestStack');

    new VerifySesDomain(stack, 'VerifyExampleDomain', {
      domainName: domain,
      addTxtRecord: false,
      addMxRecord: true,
      addDkimRecords: false,
    });

    // Then
    const template = Template.fromStack(stack);
    template.resourceCountIs('Custom::AWS', 3);
    template.resourceCountIs('AWS::SNS::Topic', 1);
    template.resourceCountIs('AWS::Route53::RecordSet', 1);
    template.hasResourceProperties('AWS::Route53::RecordSet', {
      Type: 'MX',
      Name: zoneName,
    },
    );
  });

  it('ensure dkim records are added', () => {
    const app = new cdk.App();
    const stack = new cdk.Stack(app, 'TestStack');

    new VerifySesDomain(stack, 'VerifyExampleDomain', {
      domainName: domain,
      addTxtRecord: false,
      addMxRecord: false,
      addDkimRecords: true,
    });

    // Then
    const template = Template.fromStack(stack);
    template.resourceCountIs('Custom::AWS', 4);
    template.resourceCountIs('AWS::SNS::Topic', 1);
    template.resourceCountIs('AWS::Route53::RecordSet', 3);

    const c = new Capture();
    template.hasResourceProperties('AWS::Route53::RecordSet', {
      Type: 'CNAME',
      Name: {
        'Fn::Join': ['', c],
      },
    },
    );

    expect(c.asArray()[1]).toContain('._domainkey.' + zoneName);
  });

  it('ensure custom topic is used', () => {
    const app = new cdk.App();
    const stack = new cdk.Stack(app, 'TestStack');

    const topic = new Topic(stack, 'ExistingTopic', {
      topicName: 'existing-topic-name',
    });

    new VerifySesDomain(stack, 'VerifyExampleDomain', {
      domainName: domain,
      notificationTopic: topic,
      addTxtRecord: false,
      addMxRecord: false,
      addDkimRecords: true,
    });

    // Then
    const template = Template.fromStack(stack);
    template.resourceCountIs('Custom::AWS', 4);
    template.resourceCountIs('AWS::SNS::Topic', 1);
    template.hasResourceProperties('AWS::SNS::Topic', {
      TopicName: 'existing-topic-name',
    });
  });

  it('ensure custom notification types are used', () => {
    const app = new cdk.App();
    const stack = new cdk.Stack(app, 'TestStack');

    new VerifySesDomain(stack, 'VerifyExampleDomain', {
      domainName: domain,
      notificationTypes: ['Bounce'],
      addTxtRecord: false,
      addMxRecord: false,
      addDkimRecords: true,
    });

    // Then
    const template = Template.fromStack(stack);
    template.resourceCountIs('Custom::AWS', 3);

    // since we only want to have a notification topic for Bounce, we only expect one custom resource to set a notification topic
    const c = new Capture();
    template.hasResourceProperties('Custom::AWS', {
      ServiceToken: Match.anyValue(),
      Create: Match.objectLike({
        'Fn::Join': Match.arrayWith([
          '',
          [
            c,
            Match.anyValue(),
            Match.anyValue(),
          ],
        ]),
      }),
      InstallLatestAwsSdk: true,
    });

    // now check that the custom resource is configured for 'Bounce'
    expect(c.asString()).toMatch(`{\"service\":\"SES\",\"action\":\"setIdentityNotificationTopic\",\"parameters\":{\"Identity\":\"${domain}\",\"NotificationType\":\"Bounce\"`);
  });

  it('ensure include headers enabled', () => {
    const app = new cdk.App();
    const stack = new cdk.Stack(app, 'TestStack');

    new VerifySesDomain(stack, 'VerifyExampleDomain', {
      domainName: domain,
      notificationTypes: ['Bounce'],
      includeOriginalHeaders: true,
      addTxtRecord: false,
      addMxRecord: false,
      addDkimRecords: true,
    });

    // Then
    const template = Template.fromStack(stack);
    template.resourceCountIs('Custom::AWS', 4);

    // headers are enabled for Bounce topic
    template.hasResourceProperties('Custom::AWS', {
      ServiceToken: Match.anyValue(),
      Create: Match.serializedJson(Match.objectLike({
        service: 'SES',
        action: 'setIdentityHeadersInNotificationsEnabled',
        parameters: {
          Enabled: true,
          Identity: domain,
          NotificationType: 'Bounce',
        },
      })),
      InstallLatestAwsSdk: true,
    });
  });

  it('ensure matches snapshot', () => {
    const app = new cdk.App();
    const stack = new cdk.Stack(app, 'TestStack');

    new VerifySesDomain(stack, 'VerifyExampleDomain', {
      domainName: domain,
      notificationTypes: ['Bounce'],
      includeOriginalHeaders: true,
      addTxtRecord: false,
      addMxRecord: false,
      addDkimRecords: true,
    });

    // Then
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
