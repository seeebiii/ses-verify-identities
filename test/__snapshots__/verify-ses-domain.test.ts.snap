// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SES domain verification ensure matches snapshot 1`] = `
Object {
  "Outputs": Object {
    "VerifyExampleDomainsubdomainexampleorgSesNotificationTopicBA54C794": Object {
      "Description": "SES notification topic for sub-domain.example.org",
      "Value": Object {
        "Ref": "VerifyExampleDomainSesNotificationTopicC05E5B2B",
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": Object {
      "DependsOn": Array [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "11d371859e49f3b1a49896583254890801a40beada56e4c84409a179a221e2e2.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "VerifyExampleDomainAddBounceTopicHeaderssubdomainexampleorg863B3F6D": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "VerifyExampleDomainAddBounceTopicsubdomainexampleorgCustomResourcePolicy11D5C18D",
        "VerifyExampleDomainAddBounceTopicsubdomainexampleorgD92CBEF0",
        "VerifyExampleDomainAddBounceTopicHeaderssubdomainexampleorgCustomResourcePolicyEFD7055E",
      ],
      "Properties": Object {
        "Create": "{\\"service\\":\\"SES\\",\\"action\\":\\"setIdentityHeadersInNotificationsEnabled\\",\\"parameters\\":{\\"Enabled\\":true,\\"Identity\\":\\"sub-domain.example.org\\",\\"NotificationType\\":\\"Bounce\\"},\\"physicalResourceId\\":{\\"id\\":\\"sub-domain.example.org-set-Bounce-topic-headers\\"}}",
        "InstallLatestAwsSdk": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "VerifyExampleDomainAddBounceTopicHeaderssubdomainexampleorgCustomResourcePolicyEFD7055E": Object {
      "DependsOn": Array [
        "VerifyExampleDomainAddBounceTopicsubdomainexampleorgCustomResourcePolicy11D5C18D",
        "VerifyExampleDomainAddBounceTopicsubdomainexampleorgD92CBEF0",
      ],
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ses:SetIdentityHeadersInNotificationsEnabled",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "VerifyExampleDomainAddBounceTopicHeaderssubdomainexampleorgCustomResourcePolicyEFD7055E",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "VerifyExampleDomainAddBounceTopicsubdomainexampleorgCustomResourcePolicy11D5C18D": Object {
      "DependsOn": Array [
        "VerifyExampleDomainSesNotificationTopicC05E5B2B",
      ],
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ses:SetIdentityNotificationTopic",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "VerifyExampleDomainAddBounceTopicsubdomainexampleorgCustomResourcePolicy11D5C18D",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "VerifyExampleDomainAddBounceTopicsubdomainexampleorgD92CBEF0": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "VerifyExampleDomainAddBounceTopicsubdomainexampleorgCustomResourcePolicy11D5C18D",
        "VerifyExampleDomainSesNotificationTopicC05E5B2B",
      ],
      "Properties": Object {
        "Create": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"service\\":\\"SES\\",\\"action\\":\\"setIdentityNotificationTopic\\",\\"parameters\\":{\\"Identity\\":\\"sub-domain.example.org\\",\\"NotificationType\\":\\"Bounce\\",\\"SnsTopic\\":\\"",
              Object {
                "Ref": "VerifyExampleDomainSesNotificationTopicC05E5B2B",
              },
              "\\"},\\"physicalResourceId\\":{\\"id\\":\\"sub-domain.example.org-set-Bounce-topic\\"}}",
            ],
          ],
        },
        "InstallLatestAwsSdk": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "VerifyExampleDomainSesDkimVerificationRecord0BDA51882": Object {
      "DependsOn": Array [
        "VerifyExampleDomainVerifyDomainDkimCustomResourcePolicy9E8CDFA0",
        "VerifyExampleDomainVerifyDomainDkim7F386E74",
      ],
      "Properties": Object {
        "Name": Object {
          "Fn::Join": Array [
            "",
            Array [
              Object {
                "Fn::GetAtt": Array [
                  "VerifyExampleDomainVerifyDomainDkim7F386E74",
                  "DkimTokens.0",
                ],
              },
              "._domainkey.sub-domain.example.org.",
            ],
          ],
        },
        "ResourceRecords": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::GetAtt": Array [
                    "VerifyExampleDomainVerifyDomainDkim7F386E74",
                    "DkimTokens.0",
                  ],
                },
                ".dkim.amazonses.com",
              ],
            ],
          },
        ],
        "TTL": "1800",
        "Type": "CNAME",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "VerifyExampleDomainSesDkimVerificationRecord15BC3AC32": Object {
      "DependsOn": Array [
        "VerifyExampleDomainVerifyDomainDkimCustomResourcePolicy9E8CDFA0",
        "VerifyExampleDomainVerifyDomainDkim7F386E74",
      ],
      "Properties": Object {
        "Name": Object {
          "Fn::Join": Array [
            "",
            Array [
              Object {
                "Fn::GetAtt": Array [
                  "VerifyExampleDomainVerifyDomainDkim7F386E74",
                  "DkimTokens.1",
                ],
              },
              "._domainkey.sub-domain.example.org.",
            ],
          ],
        },
        "ResourceRecords": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::GetAtt": Array [
                    "VerifyExampleDomainVerifyDomainDkim7F386E74",
                    "DkimTokens.1",
                  ],
                },
                ".dkim.amazonses.com",
              ],
            ],
          },
        ],
        "TTL": "1800",
        "Type": "CNAME",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "VerifyExampleDomainSesDkimVerificationRecord2FD3AE34A": Object {
      "DependsOn": Array [
        "VerifyExampleDomainVerifyDomainDkimCustomResourcePolicy9E8CDFA0",
        "VerifyExampleDomainVerifyDomainDkim7F386E74",
      ],
      "Properties": Object {
        "Name": Object {
          "Fn::Join": Array [
            "",
            Array [
              Object {
                "Fn::GetAtt": Array [
                  "VerifyExampleDomainVerifyDomainDkim7F386E74",
                  "DkimTokens.2",
                ],
              },
              "._domainkey.sub-domain.example.org.",
            ],
          ],
        },
        "ResourceRecords": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::GetAtt": Array [
                    "VerifyExampleDomainVerifyDomainDkim7F386E74",
                    "DkimTokens.2",
                  ],
                },
                ".dkim.amazonses.com",
              ],
            ],
          },
        ],
        "TTL": "1800",
        "Type": "CNAME",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "VerifyExampleDomainSesNotificationTopicC05E5B2B": Object {
      "DependsOn": Array [
        "VerifyExampleDomainVerifyDomainIdentityCustomResourcePolicy4DBB7566",
        "VerifyExampleDomainVerifyDomainIdentity646566FF",
      ],
      "Type": "AWS::SNS::Topic",
    },
    "VerifyExampleDomainVerifyDomainDkim7F386E74": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "VerifyExampleDomainVerifyDomainDkimCustomResourcePolicy9E8CDFA0",
        "VerifyExampleDomainVerifyDomainIdentityCustomResourcePolicy4DBB7566",
        "VerifyExampleDomainVerifyDomainIdentity646566FF",
      ],
      "Properties": Object {
        "Create": "{\\"service\\":\\"SES\\",\\"action\\":\\"verifyDomainDkim\\",\\"parameters\\":{\\"Domain\\":\\"sub-domain.example.org\\"},\\"physicalResourceId\\":{\\"id\\":\\"sub-domain.example.org-verify-domain-dkim\\"}}",
        "InstallLatestAwsSdk": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": "{\\"service\\":\\"SES\\",\\"action\\":\\"verifyDomainDkim\\",\\"parameters\\":{\\"Domain\\":\\"sub-domain.example.org\\"},\\"physicalResourceId\\":{\\"id\\":\\"sub-domain.example.org-verify-domain-dkim\\"}}",
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "VerifyExampleDomainVerifyDomainDkimCustomResourcePolicy9E8CDFA0": Object {
      "DependsOn": Array [
        "VerifyExampleDomainVerifyDomainIdentityCustomResourcePolicy4DBB7566",
        "VerifyExampleDomainVerifyDomainIdentity646566FF",
      ],
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ses:VerifyDomainDkim",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "VerifyExampleDomainVerifyDomainDkimCustomResourcePolicy9E8CDFA0",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "VerifyExampleDomainVerifyDomainIdentity646566FF": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "VerifyExampleDomainVerifyDomainIdentityCustomResourcePolicy4DBB7566",
      ],
      "Properties": Object {
        "Create": "{\\"service\\":\\"SES\\",\\"action\\":\\"verifyDomainIdentity\\",\\"parameters\\":{\\"Domain\\":\\"sub-domain.example.org\\"},\\"physicalResourceId\\":{\\"responsePath\\":\\"VerificationToken\\"}}",
        "Delete": "{\\"service\\":\\"SES\\",\\"action\\":\\"deleteIdentity\\",\\"parameters\\":{\\"Identity\\":\\"sub-domain.example.org\\"}}",
        "InstallLatestAwsSdk": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": "{\\"service\\":\\"SES\\",\\"action\\":\\"verifyDomainIdentity\\",\\"parameters\\":{\\"Domain\\":\\"sub-domain.example.org\\"},\\"physicalResourceId\\":{\\"responsePath\\":\\"VerificationToken\\"}}",
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "VerifyExampleDomainVerifyDomainIdentityCustomResourcePolicy4DBB7566": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ses:VerifyDomainIdentity",
                "ses:DeleteIdentity",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "VerifyExampleDomainVerifyDomainIdentityCustomResourcePolicy4DBB7566",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
